params.go

```diff
@@ -84,7 +84,7 @@ func cGCMParams(p *GCMParams) []byte {
 	p.Free()
 	p.arena = arena
 	p.params = &params
-	return C.GoBytes(unsafe.Pointer(&params), C.int(unsafe.Sizeof(params)))
+	return memBytes(unsafe.Pointer(&params), unsafe.Sizeof(params))
```

Comment:
But this memory isn't allocated by the Go runtime? So how/why can we let it clean it up?

Reply:
The `params` struct is allocated by the Go runtime in line 69. The only C memory that is allocated here is the memory pointed to by the `pIv` and `pAAD` pointers, which are owned by the `arena` list.

Reply:
Maybe this is the confusion: While `C.CString()` and `C.CBytes()` both allocate C memory, `C.GoString()` and `C.GoBytes()` don't free any C memory. If that would be the case, it would crash before, because it would try to free Go allocated memory. All four functions do copy the data, though. So my change only skips the copy of the Go allocated memory. It does not change anything else.

Reply:
ah, thanks. This may/might also impact https://github.com/miekg/pkcs11/issues/103

Are you able to test on ARM? I've got severals PIs here, but none of the is operational

Reply:
My Pi is in use. But there should be ARM containers/VMs available... Let me check...

Reply:
I just ran the tests in Docker for Mac which includes QEMU Arm Emulation:
<details><summary>ARM Test</summary>
<pre><code>~/pkcs11 # go test -v
=== RUN   TestConstCouunt
--- PASS: TestConstCouunt (0.09s)
=== RUN   TestPSSParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestPSSParams (0.01s)
=== RUN   TestOAEPParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestOAEPParams (0.00s)
=== RUN   TestGCMParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestGCMParams (0.00s)
=== RUN   TestSetenv
    pkcs11_test.go:50: Failed to init pkcs11
--- FAIL: TestSetenv (0.00s)
=== RUN   TestInitialize
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestInitialize (0.00s)
=== RUN   TestNew
--- PASS: TestNew (0.00s)
=== RUN   TestGetInfo
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestGetInfo (0.00s)
=== RUN   TestFindObject
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestFindObject (0.00s)
=== RUN   TestGetAttributeValue
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestGetAttributeValue (0.00s)
=== RUN   TestDigest
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestDigest (0.00s)
=== RUN   TestDigestUpdate
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestDigestUpdate (0.00s)
=== RUN   TestGenerateKeyPair
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestGenerateKeyPair (0.00s)
=== RUN   TestSign
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestSign (0.00s)
=== RUN   TestDestroyObject
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestDestroyObject (0.00s)
=== RUN   TestSymmetricEncryption
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:36: Failed to init lib
--- FAIL: TestSymmetricEncryption (0.00s)
=== RUN   ExampleCtx_Sign
2022/01/20 16:04:55 Failed to init lib
exit status 1
FAIL	github.com/miekg/pkcs11	0.269s
~/pkcs11 # apk add softhsm
(1/9) Installing ncurses-terminfo-base (6.3_p20211120-r0)
(2/9) Installing ncurses-libs (6.3_p20211120-r0)
(3/9) Installing readline (8.1.1-r0)
(4/9) Installing sqlite (3.36.0-r0)
(5/9) Installing libbz2 (1.0.8-r1)
(6/9) Installing xz-libs (5.2.5-r0)
(7/9) Installing sqlite-libs (3.36.0-r0)
(8/9) Installing botan-libs (2.18.1-r4)
(9/9) Installing softhsm (2.6.1-r2)
Executing busybox-1.34.1-r3.trigger
OK: 515 MiB in 44 packages
~/pkcs11 # go test -v
=== RUN   TestConstCouunt
--- PASS: TestConstCouunt (0.09s)
=== RUN   TestPSSParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestPSSParams (0.52s)
=== RUN   TestOAEPParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestOAEPParams (0.54s)
=== RUN   TestGCMParams
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestGCMParams (0.03s)
=== RUN   TestSetenv
--- PASS: TestSetenv (0.00s)
=== RUN   TestInitialize
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestInitialize (0.00s)
=== RUN   TestNew
--- PASS: TestNew (0.00s)
=== RUN   TestGetInfo
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:110: {CryptokiVersion:{Major:2 Minor:40} ManufacturerID:SoftHSM Flags:0 LibraryDescription:Implementation of PKCS11 LibraryVersion:{Major:2 Minor:6}}
--- PASS: TestGetInfo (0.02s)
=== RUN   TestFindObject
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestFindObject (0.38s)
=== RUN   TestGetAttributeValue
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
    pkcs11_test.go:157: attr 0, type 290, valuelen 3
    pkcs11_test.go:157: attr 1, type 289, valuelen 8
    pkcs11_test.go:157: attr 2, type 288, valuelen 256
    pkcs11_test.go:161: modulus 23177777739565763036282288451612277215094384267580649141662939060143653520262815874118203995592856707364133701746712231446013495003062418685260033536897721621719332973876868208594342698847009386797801425213144004540616330415527202157984084012213855606210350141680422980461205958516644774066799652962912949996909837145873324004715098082409084003064991158954258350469636813350730092911010114461081646533062309803387303320248286189186258025689039259133805988684837466194122767775234756766518600432835863922488209702742743436552216975317554938048854637024088066334038385870014335695062757114883232386654747089000511233519
    pkcs11_test.go:157: attr 3, type 3, valuelen 17
--- PASS: TestGetAttributeValue (0.23s)
=== RUN   TestDigest
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
=== RUN   TestDigest/Simple
=== RUN   TestDigest/Empty
--- PASS: TestDigest (0.02s)
    --- PASS: TestDigest/Simple (0.00s)
    --- PASS: TestDigest/Empty (0.00s)
=== RUN   TestDigestUpdate
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
=== RUN   TestDigestUpdate/Simple
=== RUN   TestDigestUpdate/Empty
--- PASS: TestDigestUpdate (0.01s)
    --- PASS: TestDigestUpdate/Simple (0.00s)
    --- PASS: TestDigestUpdate/Empty (0.00s)
=== RUN   TestGenerateKeyPair
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestGenerateKeyPair (0.47s)
=== RUN   TestSign
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestSign (0.92s)
=== RUN   TestDestroyObject
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
--- PASS: TestDestroyObject (0.41s)
=== RUN   TestSymmetricEncryption
    pkcs11_test.go:33: loading /usr/lib/softhsm/libsofthsm2.so
=== RUN   TestSymmetricEncryption/Encrypt
=== RUN   TestSymmetricEncryption/Encrypt/ECB
=== RUN   TestSymmetricEncryption/Encrypt/CBC
=== RUN   TestSymmetricEncryption/Encrypt/CBC-PAD
=== RUN   TestSymmetricEncryption/EncryptUpdate
=== RUN   TestSymmetricEncryption/EncryptUpdate/ECB
=== RUN   TestSymmetricEncryption/EncryptUpdate/CBC
=== RUN   TestSymmetricEncryption/EncryptUpdate/CBC-PAD
--- PASS: TestSymmetricEncryption (0.04s)
    --- PASS: TestSymmetricEncryption/Encrypt (0.01s)
        --- PASS: TestSymmetricEncryption/Encrypt/ECB (0.00s)
        --- PASS: TestSymmetricEncryption/Encrypt/CBC (0.00s)
        --- PASS: TestSymmetricEncryption/Encrypt/CBC-PAD (0.00s)
    --- PASS: TestSymmetricEncryption/EncryptUpdate (0.01s)
        --- PASS: TestSymmetricEncryption/EncryptUpdate/ECB (0.01s)
        --- PASS: TestSymmetricEncryption/EncryptUpdate/CBC (0.00s)
        --- PASS: TestSymmetricEncryption/EncryptUpdate/CBC-PAD (0.00s)
=== RUN   ExampleCtx_Sign
--- PASS: ExampleCtx_Sign (0.37s)
PASS
ok  	github.com/miekg/pkcs11	4.188s
~/pkcs11 # go version
go version go1.17.4 linux/arm64
~/pkcs11 # git log --oneline | head -1
02b3205 Remove unnecessary memory copies by C.GoBytes() calls
~/pkcs11 # 
</code></pre>
</details>

Reply:
Not sure it's related to #103. The `C.free(unsafe.Pointer(sig))` looks legit to me.

Reply:
To double check, I just ran the tests of my PR on an AWS arm64 instance, same results.